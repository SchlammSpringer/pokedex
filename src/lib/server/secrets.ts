export const secret =
  'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imh5a3J0eGtkcnJnc2FianN1YXZ4Iiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTY4MDY0MDk4MiwiZXhwIjoxOTk2MjE2OTgyfQ.glF-TpZ4tn6DPIhw9fXcn49JFwAV0ooQ68p2KpEt65c'
