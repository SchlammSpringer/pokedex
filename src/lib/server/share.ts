import type {
  Flavor,
  LanguageName,
  Pokemon,
  PokemonFromApi,
  SpeciesFromApi,
  Type
} from '$lib/types'

export const fetchPokemon = async (pokeUrl: URL) => {
  const pokemonResponse: Response = await fetch(pokeUrl, {
    headers: {
      Accept: 'application/json'
    }
  })
  const pokemonFromApi: PokemonFromApi = await pokemonResponse.json()
  const speciesResponse: Response = await fetch(pokemonFromApi.species.url, {
    headers: {
      Accept: 'application/json'
    }
  })
  const species: SpeciesFromApi = await speciesResponse.json()
  const pokemon: Pokemon = {
    pokedex: pokemonFromApi.id,
    name: pokemonFromApi.name,
    description: extractFlavorTexts(species).join('\n'),
    types: pokemonFromApi.types.map((type: Type) => type.type.name),
    color: species.color.name,
    germanName: species.names
      .filter((name: LanguageName) => name.language.name === 'de')
      .map((name: LanguageName) => name.name)
      .join('\n'),
    habitat: species.habitat.name
  }
  return pokemon
}

const extractFlavorTexts = ({ flavor_text_entries }: { flavor_text_entries: Flavor[] }) => {
  const allTexts = flavor_text_entries
    .filter((flavor: Flavor) => flavor.language.name === 'de')
    .map((flavor: Flavor) => flavor.flavor_text.replace(/[\r\n\f]/gm, ' '))
  return [...new Set(allTexts)]
}
