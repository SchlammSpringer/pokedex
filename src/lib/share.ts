export const spriteUrl = 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/'
export const spriteBackUrl =
  'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/'

export const spriteShinyUrl =
  'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/shiny/'
export const spriteShinyBackUrl =
  'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/shiny/'

export const officialArtworkUrl = './assets/official-artwork/'
export const officialShinyArtworkUrl = './assets/official-artwork/shiny/'
