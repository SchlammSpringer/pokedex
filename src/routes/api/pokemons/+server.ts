import type { RequestHandler } from '@sveltejs/kit'
import { json } from '@sveltejs/kit'
import type { Pokemons, PokemonShort } from '../../../types'

const lastPart = (href: string) => {
  const segments = new URL(href).pathname.split('/')
  return segments.pop() || segments.pop()
}

const toPokemonShort = () => (pokemon: { url: string; name: string }) => ({
  pokedex: lastPart(pokemon.url),
  name: pokemon.name
})

const fetchOriginalPokemons = async () => {
  const response = await fetch('https://pokeapi.co/api/v2/pokemon?limit=151')
  return await response.json()
}

export const GET = (async () => {
  const originalPokemons = await fetchOriginalPokemons()
  const pokemonShorts: PokemonShort[] = originalPokemons.results.map(toPokemonShort())
  return json(<Pokemons>{ pokemons: pokemonShorts })
}) satisfies RequestHandler
