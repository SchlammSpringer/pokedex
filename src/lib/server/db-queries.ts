import type { Pokemon } from '$lib/types'
import { supabase } from './supabase'

export const selectAllPokemons = () =>
  supabase.from('Pokemons').select('pokemon', { count: 'exact' }).order('id')

export const insertAllPokemons = (pokemon: Pokemon) =>
  supabase.from('Pokemons').insert([{ id: pokemon.pokedex, pokemon: pokemon }])

export const findPokemonByName = (name: string) =>
  supabase.from('Pokemons').select('pokemon').eq('pokemon->>name', name)

export const updatePokemon = (pokemon: Pokemon) =>
  supabase.from('Pokemons').update({ pokemon: pokemon }).eq('id', pokemon.pokedex).select()
