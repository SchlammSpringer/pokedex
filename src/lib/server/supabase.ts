import { createClient } from '@supabase/supabase-js'
import { secret } from '$lib/server/secrets'
import type { Database } from '$lib/db/types'

const supabaseUrl = 'https://hykrtxkdrrgsabjsuavx.supabase.co'
const supabaseKey = secret
export const supabase = createClient<Database>(supabaseUrl, supabaseKey)
