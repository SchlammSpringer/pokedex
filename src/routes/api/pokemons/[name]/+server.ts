import { BYPASS_TOKEN } from '$env/static/private'
import { findPokemonByName, updatePokemon } from '$lib/server/db-queries'
import { validatePokemon } from '$lib/server/share'
import { type Pokemon } from '$lib/types'
import type { RequestEvent, RequestHandler } from '@sveltejs/kit'
import { error, json, error as svelteError } from '@sveltejs/kit'

export const GET = (async ({ params }) => {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const { data } = await findPokemonByName(params.name!)
  if (!data || data?.length === 0) throw error(404, 'not found')
  return json(data[0].pokemon)
}) satisfies RequestHandler

export const PUT = async (event: RequestEvent) => {
  const body: Pokemon = (await event.request.json()) as Pokemon

  await validatePokemon(body)

  const { data } = await updatePokemon(body)
  if (data) {
    await event.fetch('/api/pokemons/' + event.params.name, {
      headers: { 'x-prerender-revalidate': BYPASS_TOKEN },
      method: 'HEAD'
    })
    return json(data[0].pokemon)
  } else {
    throw svelteError(400, 'validation errors: ' + JSON.stringify(form.errors, null, 2))
  }
}
