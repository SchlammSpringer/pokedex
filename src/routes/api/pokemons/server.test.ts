import type { RequestEvent } from '@sveltejs/kit'
import { json } from '@sveltejs/kit'
import { type Mock, vi } from 'vitest'
import { mock } from 'vitest-mock-extended'
import { GET, config } from './+server'
import createFetchMock from 'vitest-fetch-mock'
import { selectAllPokemons } from '$lib/server/db-queries'
import type { Pokemon } from '$lib/types'
import { fetchPokemon } from '$lib/server/share'

const fetchMocker = createFetchMock(vi)
fetchMocker.enableMocks()

describe('Pokemon list API', () => {
  afterEach(() => {
    vi.restoreAllMocks()
  })

  vi.mock('@sveltejs/kit')
  vi.mock('$lib/server/db-queries')
  vi.mock('$lib/server/share')
  const request = mock<RequestEvent>()

  const pokemonsFromApi = {
    results: [{ url: 'https://pokemon/1' }, { url: 'https://pokemon/2' }]
  }

  const pokemons: Pokemon[] = [
    {
      pokedex: 1,
      name: 'bulbasaur',
      description: 'Dieses Pokémon trägt von Geburt an einen Samen auf dem Rücken,',
      types: ['grass', 'poison'],
      color: 'green',
      germanName: 'Bisasam',
      habitat: 'grassland'
    },
    {
      pokedex: 25,
      name: 'pikachu',
      description: 'Es hat kleine Backentaschen',
      types: ['electric'],
      color: 'yellow',
      germanName: 'Pikachu',
      habitat: 'forest'
    }
  ]

  const mockedFindPokemons = selectAllPokemons as Mock
  const mockedFetchPokemon = fetchPokemon as Mock

  interface Response {
    pokemons: Pokemon[]
    header: Record<string, string>
  }

  const convertResponse = () => (pokemons: Pokemon[], header: Record<string, string>) =>
    ({
      pokemons,
      header
    } satisfies Response)

  const header = {
    headers: { 'Cache-Control': 'public, max-age=0, s-maxage=86400' }
  }
  it('check vercel config', () => {
    expect(config).toMatchSnapshot()
  })

  describe('GET', () => {
    it('from pokeapi', async () => {
      mockedFindPokemons.mockResolvedValue({ data: [] })
      fetchMocker.mockIf('https://pokeapi.co/api/v2/pokemon?limit=151', (req) =>
        JSON.stringify(pokemonsFromApi)
      )
      mockedFetchPokemon.mockResolvedValueOnce(pokemons[0])
      mockedFetchPokemon.mockResolvedValueOnce(pokemons[1])
      const mockedJson = json as Mock
      mockedJson.mockImplementation(convertResponse())

      const response = await GET()

      expect(response).toStrictEqual({ pokemons, header })
    })

    it('from db ', async () => {
      mockedFindPokemons.mockResolvedValue({
        data: [{ pokemon: pokemons[0] }, { pokemon: pokemons[1] }]
      })
      const mockedJson = json as Mock
      mockedJson.mockImplementation(convertResponse())

      const response = await GET()

      expect(mockedFindPokemons).toHaveBeenCalledTimes(1)
      expect(response).toStrictEqual({ pokemons, header })
    })
  })
})
