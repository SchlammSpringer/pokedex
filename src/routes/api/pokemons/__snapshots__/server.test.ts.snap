// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pokemon list API > check vercel config 1`] = `
{
  "regions": [
    "fra1",
  ],
  "runtime": "edge",
  "split": true,
}
`;
